grin/src/sync.rs

```diff
@@ -294,21 +369,17 @@ impl Syncer {
 		// both nodes share at least one common header hash in the locator
 		heights.push(0);
 
-		debug!(LOGGER, "Sync: Loc heights: {:?}", heights);
+		debug!(LOGGER, "Sync: Locator heights (from sync buddy): {:?}", heights);
 
-		// Iteratively travel the header chain back from our head and retain the
+		// Iteratively travel the header chain back from peer's head and retain the
 		// headers at the wanted heights.
-		let mut header = self.chain.get_block_header(&tip.last_block_h)?;
-		let mut locator = vec![];
-		while heights.len() > 0 {
-			if header.height == heights[0] {
-				heights = heights[1..].to_vec();
-				locator.push(header.hash());
-			}
-			if header.height > 0 {
-				header = self.chain.get_block_header(&header.previous)?;
-			}
-		}
+	let locator = heights
+		.into_iter()
+		.map(|h| self.chain.get_header_by_height(h))
```

Comment:
This still won't work for 2 reasons:

* The header isn't stored under a height under the full block is received is validated. Could be changed but would introduce some more complexity and still not solve the next point.
* If you're syncing fork A, and your buddy flips (or you have 2 alternating buddies), you'll get the wrong headers and wrong hashes.

Iterating back from the tip hash is your only safe bet here.

Reply:
Yes, was got confused again between the header chain and the 'validated' chain, makes perfect sense.